### PR TITLE
🌱  Update golangci-lint to 1.49.0

### PR DESCRIPTION
**Pull request by killianmuldoon**
Signed-off-by: killianmuldoon <kmuldoon@vmware.com>

Update golangci-lint to version 1.49. This update re-enables the revive rules that were disabled in v1.48.0

### CODE REVIEW

**Review by killianmuldoon — COMMENTED** [review]
Had to add package comments to all of our main packages - possibly we could exclude them, but I don't think it does any harm.

**Comment by killianmuldoon** [comment]
/hold 

Until https://github.com/kubernetes-sigs/cluster-api/issues/7115 is decided.

File: controlplane/kubeadm/main.go [hunk 1]
```diff
@@ -158,7 +159,8 @@ func main() {
 	if profilerAddress != "" {
 		klog.Infof("Profiler listening for requests at %s", profilerAddress)
 		go func() {
-			klog.Info(http.ListenAndServe(profilerAddress, nil))
+			srv := http.Server{Addr: profilerAddress, ReadHeaderTimeout: 2 * time.Second}
```
Comment (killianmuldoon) [inline controlplane/kubeadm/main.go:162]: This was done to address a new rule in gosec (G112) to prevent a slow loris attack. I'm happy to alternatively ignore that here as this is for debug purposes.

File: bootstrap/kubeadm/main.go [hunk 2]
```diff
@@ -151,11 +152,11 @@ func main() {
 
 	// klog.Background will automatically use the right logger.
 	ctrl.SetLogger(klog.Background())
-
 	if profilerAddress != "" {
 		klog.Infof("Profiler listening for requests at %s", profilerAddress)
 		go func() {
-			klog.Info(http.ListenAndServe(profilerAddress, nil))
+			srv := http.Server{Addr: profilerAddress, ReadHeaderTimeout: 2 * time.Second}
+			klog.Info(srv.ListenAndServe())
```
Comment (sbueringer) [inline bootstrap/kubeadm/main.go:159]: Can we change the log to do the same as for manager? e.g.
```go
			if err := srv.ListenAndServe(); err != nil {
				klog.ErrorS(err, "problem running profiler server")
			}
```

P.S. I don't know why klog is used here instead of setupLog, maybe there's a reason. (I'm fine with just keep using klog in this go routine)

File: bootstrap/kubeadm/main.go [hunk 3]
```diff
@@ -14,6 +14,7 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
+// main is the main package for Kubeadm Bootstrap.
```
Comment (sbueringer) [inline bootstrap/kubeadm/main.go:17]:
```suggestion
// main is the main package for the Kubeadm Bootstrap controller.
```
nit (please also check the other controller package comments)

**Review by sbueringer — COMMENTED** [review]
just nits

Thx for taking care of this

File: bootstrap/kubeadm/main.go (hunk 2 above)
Comment (killianmuldoon) [reply bootstrap/kubeadm/main.go:159]: To use Setuplog we have to do 
```suggestion
	                setupLog.Info(srv.ListenAndServe().Error())
```

I think it's fine to leave as is though.

**Comment by killianmuldoon** [comment]
I squashed the commits so this sing commit now contains the removal of deprecated linters.

File: bootstrap/kubeadm/main.go [hunk 4]
```diff
@@ -151,11 +152,13 @@ func main() {
...
 	// klog.Background will automatically use the right logger.
 	ctrl.SetLogger(klog.Background())
-
 	if profilerAddress != "" {
 		klog.Infof("Profiler listening for requests at %s", profilerAddress)
 		go func() {
-			klog.Info(http.ListenAndServe(profilerAddress, nil))
+			srv := http.Server{Addr: profilerAddress, ReadHeaderTimeout: 2 * time.Second}
+			if err := srv.ListenAndServe(); err != nil {
+				setupLog.Error(err, "problem running profiler server")
```
Comment (sbueringer) [inline bootstrap/kubeadm/main.go:160]: Given that we switched here to setupLog, let's please also switch in l.156

**Review by sbueringer — COMMENTED** [review]
Thx!

last nit

File: bootstrap/kubeadm/main.go [hunk 5]
```diff
@@ -151,11 +152,13 @@ func main() {
 
 	// klog.Background will automatically use the right logger.
 	ctrl.SetLogger(klog.Background())
-
 	if profilerAddress != "" {
-		klog.Infof("Profiler listening for requests at %s", profilerAddress)
+		setupLog.Info("Profiler listening for requests at %s", profilerAddress)
```
Comment (sbueringer) [inline bootstrap/kubeadm/main.go:156]: setupLog.Info has this signature:
```go
func (l Logger) Info(msg string, keysAndValues ...interface{}) {
```

Afaik this log line would panic because of uneven keysAndValues count

Comment (killianmuldoon) [reply bootstrap/kubeadm/main.go]: Thanks for catching that! Replaced now.

**Comment by sbueringer** [comment]
Thx!
/lgtm

File: .golangci.yml [hunk 6]
```diff
@@ -251,15 +238,6 @@ issues:
...
-    - ifshort
-    text:  "variable 'isDeleteNodeAllowed' is only used in the if-statement.*"
-    path: ^internal/controllers/machine/machine_controller\.go$
-  - linters:
-    - ifshort
-    text: "variable 'kcpMachinesWithErrors' is only used in the if-statement.*"
-    path: ^controlplane/kubeadm/internal/workload_cluster_conditions\.go$
   # We don't care about defer in for loops in test files.
   - linters:
     - gocritic
```
Comment (oscr) [inline .golangci.yml:243]: Nit: could also delete as a minor cleanup:

```
 skip-dirs:
  - third_party
```
 https://github.com/kubernetes-sigs/cluster-api/blob/3f6f56fb556cd8855b85dd9e878d5e373fc26ede/.golangci.yml#L277

File: main.go [hunk 7]
```diff
@@ -216,9 +217,12 @@ func main() {
 	ctrl.SetLogger(klog.Background())
 
 	if profilerAddress != "" {
-		klog.Infof("Profiler listening for requests at %s", profilerAddress)
+		setupLog.Info(fmt.Sprintf("Profiler listening for requests at %s", profilerAddress))
 		go func() {
-			klog.Info(http.ListenAndServe(profilerAddress, nil))
+			srv := http.Server{Addr: profilerAddress, ReadHeaderTimeout: 2 * time.Second}
```
Comment (ykakarap) [inline main.go:222]: Out of curiosity: Was this change made to fix some lint error? Seems unrelated otherwise.

**Review by ykakarap — COMMENTED** [review]
/lgtm

File: main.go (hunk 7 above)
Comment (ykakarap) [reply main.go:222]: Never mind: This was address in a resolved comment. 😆

**Review by fabriziopandini — COMMENTED** [review]
/lgtm
/approve

File: .golangci.yml (hunk 6 above)
Comment (sbueringer) [reply .golangci.yml:243]: Addressed in: https://github.com/kubernetes-sigs/cluster-api/pull/7122/files
